src/ManageCourses.Api/Services/Data/DataService.cs

```diff
@@ -30,18 +32,124 @@ public DataService(IManageCoursesDbContext context, IDataHelper dataHelper, ILog
         /// <param name="payload">Holds all the data entities that need to be imported</param>
         public void ProcessUcasPayload(UcasPayload payload)
         {
-            ResetUcasSchema();
+            var allInstitutions = _context.UcasInstitutions.ToList();
+            
+            _logger.LogWarning("Beginning UCAS import");
+            _logger.LogInformation($"Upserting {allInstitutions.Count()} institutions");
+            int processed = 0;
+            foreach (var inst in allInstitutions)
+            {
+                var deleted = AttemptTransaction(inst.InstCode, "delete for replace", () => DeleteForInstitution(inst.InstCode));
+                var added = AttemptTransaction(inst.InstCode, "insert", () => AddForInstitution(inst.InstCode, payload));
```

Comment:
Why wouldn't we just wrap the delete+add in a single transaction and do away with all the EF state tracking and attempted rollbacks?

Reply:
Answer: was already attempted.